atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/keyvalue/api/cache/LockWatchValueCacheImpl.java

```diff
@@ -0,0 +1,163 @@
+        commitUpdate.accept(new Visitor<Void>() {
+            @Override
+            public Void invalidateAll() {
+                // If this is an election, we should throw. If not and we are not a serialisable transaction, we are
+                // *technically* ok to go on here. However, we cannot determine which of the two we are in, so we
+                // throw anyway.
+                throw new TransactionLockWatchFailedException("A Timelock leader election has occurred between the "
+                        + "start and commit time of this transaction, and thus all events have been invalidated. This"
+                        + " transaction will be retried");
+            }
+
+            @Override
+            public Void invalidateSome(Set<LockDescriptor> invalidatedLocks) {
+                Set<CellReference> invalidatedCells = invalidatedLocks.stream()
+                        .flatMap(LockWatchValueCacheImpl::extractTableAndCell)
+                        .collect(Collectors.toSet());
+                KeyedStream.stream(digest.loadedValues())
+                        .filterKeys(cellReference -> !invalidatedCells.contains(cellReference))
+                        .forEach(valueStore::putValue);
+                return null;
+            }
+        });
+        snapshotStore.removeTimestamp(StartTimestamp.of(startTs));
+    }
+
+    @Override
+    public TransactionScopedCache createTransactionScopedCache(long startTs) {
+        // todo(jshah): implement
+        throw new UnsupportedOperationException("Not yet implemented");
+    }
+
+    /**
+     * In order to maintain the necessary invariants, we need to do the following:
+     *
+     *  1. For each new event, we apply it to the cache. The effects of this application is described in
+     *     {@link LockWatchValueCache}.
+     *  2. For each transaction, we must ensure that we store a snapshot of the cache at the sequence corresponding
+     *     to the transaction's start timestamp. Note that not every sequence will have a corresponding timestamp, so we
+     *     don't bother storing a snapshot for those sequences. Also note that we know that each call here will only
+     *     ever have new events, and that consecutive calls to this method will *always* have increasing sequences
+     *     (without this last guarantee, we'd need to store snapshots for all sequences).
+     */
+    private void updateStores(TransactionsLockWatchUpdate updateForTransactions) {
+        Multimap<Sequence, StartTimestamp> reversedMap = createSequenceTimestampMultimap(updateForTransactions);
+        updateForTransactions.events().forEach(event -> {
+            valueStore.applyEvent(event);
+            Sequence sequence = Sequence.of(event.sequence());
+            reversedMap
+                    .get(sequence)
+                    .forEach(timestamp -> snapshotStore.storeSnapshot(sequence, timestamp, valueStore.getSnapshot()));
```

Comment:
The potential allocations here look scary; I _think_ this is OK because we have vavr, but we should confirm this is fine.

Reply:
Internally, we tend to have often >100 snapshots of the cache at any one time. I expect that this is fine, but I'll just dig in vavr's docs to confirm.

Reply:
Had a look, it definitely states that this is the kind of pattern that can be handled. Also, the difference in interface between vavr's map and java's basic map means that it should be quite hard for us to accidentally switch these.